### DISCUSSION TITLE
subscription-ci Pipeline no changes detected

### DISCUSSION THREAD

**Discussion by theMK2000** [Azure/CanadaPubSecALZ#360] [Q&A] [answered]
I am not sure if anyone can assist with this but when I run the subscriptions-ci pipeline in azure ado it succeeds but it is saying under the trigger subscription deployment

No subscription configuration changes detected in current commit ()

So none of the code changes appear in the subscription I have spent time trying to resolve this issue. Can anyone advise me on troubleshooting this? Let me know if you need any more information.  I am sure this is a simple error on my side.

**Comment by skeeler** [comment]
@theMK2000  the subscriptions pipeline can be triggered or run manually.

When triggered (not the default state in the `subscriptions.yml` pipeline definition) it will examine the files in the triggering commit and run the pipeline (as parallel jobs) on any included subscription configuration files (newly added) under the `/config/subscriptions` path.

When run manually, there are two parameters requested, one of which must be supplied:

- The commit id parameter allows you to provide a previous git commit hash value, and the pipeline logic from that point is similar to the flow mentioned in the triggered case.
- The subscription ids (a comma-separated array) parameter value is the most often used option when running the pipeline manually. In this case, you need to provide the subscription GUID(s) - or enough of the prefix of each so it/they can be uniquely found under the `/config/subscriptions` folder in the Git repo. The pipeline will perform a recursive prefix matching search for each value to locate its corresponding configuration file.

If you're already doing this, then the next thing to check is ensuring that your subscription configuration files are located under the correct folder path.

For example, let's say I have a config file for the following subscription:

- My Azure DevOps organization name is: `MyAzDevOpsOrg`
- The branch I am running pipeline & accessing repo code from is: `main`
- The subscription GUID is `4f9f8765-911a-4a6d-af60-4bc0473268c0` and that uniquely identifies it across my subscriptions
- I want to use the `lz-generic-subscription` landing zone template (archetype)
- I want to place it in the `canadaeast` region
- I want to move it under a management group with the identity `LandingZonesProduction`

With the above requirements, I would need to create a subscription configuration file with the following name & location:

  `/config/subscriptions/MyAzDevOpsOrg-main/LandingZonesProduction/4f9f8765-911a-4a6d-af60-4bc0473268c0_generic-subscription_canadaeast.json`

When I run the `subscriptions-ci` pipeline (on the `main` branch), I can enter `4f9f` as the subscription id - assuming there are no other subscriptions that start with the same 4 hex digits under the `/config/subscriptions/MyAzDevOpsOrg-main` folder hierarchy.

**Reply by theMK2000** [reply]
@skeeler Thank you for your reply. Everything you have mentioned I have done I have also been able to run this on a new subscription without issue. I am currently taking one of my old subscriptions and moving it to a new management group and then running the pipeline. When the pipeline runs it detects the JSON file but for some reason does not apply any of the changes to the subscription. Please see the pipeline log file attached. Thanks again.
[sucscriptionCI.txt](https://github.com/Azure/CanadaPubSecALZ/files/10962035/sucscriptionCI.txt)

**Reply by skeeler** [reply]
@theMK2000, for existing subscription with modifications and the `subscriptions-ci` being triggered by a commit (or the commit hash value being provided as a parameter when running the pipeline manually), there is a variable named `var-TriggerSubscriptionDeployOn` located in the `/config/variables/common.yml` file that is, by default, set to `A`. What this means is only act on "Added" subscription config files. To have the subscription pipeline process both Added and Modified config files (via commit), change it to: `var-TriggerSubscriptionDeployOn: 'AM'` (to include both "Added" and "Modified" files in the commit). You can change it either in the `common.yml` file *or* in your environment YAML file (located under `/config/variables`) where any repeated variable settings are overridden (`common.yml` is loaded first, then your specific environment YAML variables config are loaded after).

This is the relevant section from your log file showing that a file was modified, but the change status flag is set to `A`:

![image](https://user-images.githubusercontent.com/755941/225155077-15f6e5ec-648a-4ec9-974e-457d9f2ec7d1.png)

**Reply by theMK2000** [reply]
Thank you now I am getting another error.

`Invoke-RestMethod: /home/vsts/work/_temp/01b210fa-a496-4239-a44c-3a4f91a414f6.ps1:16
Line |
  16 |  … $pipeline = Invoke-RestMethod -Uri $url -Headers $headers -Method Pos …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | {"$id":"1","innerException":null,"message":"Stage
     | DeploySubscriptionsStage job
     | DeploySubscriptionJob_BCFinancialServicesAuthority_main/bcfsasandbox/618eb669_d7d2_4148_bee9_d93f9e932cd2_generic_subscription_canadacentral.json has an invalid name. Valid names may only contain alphanumeric characters and '_' and may not start with a number.","typeName":"Microsoft.Azure.Pipelines.WebApi.PipelineValidationException, Microsoft.Azure.Pipelines.WebApi","typeKey":"PipelineValidationException","errorCode":0,"eventId":3000}

`

I believe this has something to do with trigger-subscriotions.yml  I will try and trouble shoot this tomorrow thank you for helping

**Reply by skeeler** [reply]
@theMK2000 sorry it's taken a while to respond. From the error message above, it appears your subscription configuration file name and the archetype/template name are both using the underscore ("_") character, but should be using the dash ("-") character instead. E.g.  `618eb669-d7d2-4148-bee9-d93f9e932cd2_generic-subscription_canadacentral.json`. The underscore is only used to separate the `subscription GUID` from the `archetype/template` from the `region/datacenter`.

**Reply by theMK2000** [reply]
@skeeler Thanks for responding I used the bellow on my file name but still get the error.

![image](https://user-images.githubusercontent.com/127534928/228069986-0769e963-f093-4e2f-bd9e-3d52197bed7a.png)

As you can see 1) the value is correct I believe but then once the file name is parsed at 2) it is all _ though the issue seems to be that it starts with a # I am not 100% sure that is the issue and not sure how to get around it do I add _ at the start?

**Reply by skeeler** [reply]
@theMK2000, I will try to reproduce today in my environment to see if it's an issue with the pipeline logic or perhaps something environmental. While I'm working on this, I'd appreciate if you could share the approximate date your code was cloned/forked from the main repository, and whether you have made any changes to the pipeline logic.

**Reply by theMK2000** [reply]
We cloned it about 3-4 months ago.  We did not make any changes to the pipeline logic.

**Comment by skeeler** [comment, marked as answer]
hi @theMK2000, I examined several places to ensure the pipeline task script logic was valid (it was). In the end, it turns out that the `Invoke-RestMethod` call is failing due to the presence of the `/` characters in the fully qualified path name of each Added or Modified subscription configuration (JSON) file. This code hasn't changed since before 2021-09-30 and was working then, so I can only assume that something in the PowerShell module containing the `Invoke-RestMethod` cmdlet has changed in the interim.

To fix the error you are encountering, you'll need to make the following changes to the `.pipelines/templates/jobs/trigger-subscriptions.yml` file and commit to your Azure DevOps repository. Following are both a visual and textual representation of the changes needed:.

First, the visual representation. The code additions are marked in green on the right hand side, and the code changes are marked in red/green with green representing final state:

![trigger-subscriptions-yml-changes](https://user-images.githubusercontent.com/755941/229221905-95bd8b83-581c-4803-9d70-2bc4d8ce613a.png)

For the textual representation, you will copy/paste the following block of code directly following line 91 in the current code:

```PowerShell
$paths = $env:SUBSCRIPTION_CHANGES -split ','
$guids = $paths -replace '.*?([0-9a-f]{8}[-]?([0-9a-f]{4}[-]?){3}[0-9a-f]{12}).*', '$1'
$changes = $guids -join ','
```

After pasting the above code block after existing line 91, you will change the `$env:SUBSCRIPTION_CHANGES` variable reference (on what is now line 98) to be: `$changes`.

I tested these changes by performing the following actions:
1. Set the `var-TriggerSubscriptionDeployOn` variable to `AM` (in `common.yml`)
2. Enabling CI `trigger` in the `subscriptions.yml` pipeline definition file
3. Adding a new subscription configuration file and monitoring its 2 resultant pipeline runs through successful completion
4. Modifying an existing subscription configuration file and monitoring its 2 resultant pipeline runs through successful completion

The results of the 2 pipeline CI triggers (each spawning a manual pipeline run) are shown below:

![image](https://user-images.githubusercontent.com/755941/229241227-7a08ad7f-a199-4eb1-a6bc-b19b5928c889.png)

These pipeline runs are displayed in reverse order of execution.

**Reply by theMK2000** [reply]
Thank you for this it still coming up with the error though the code change has updated the file name not to include /.  I will do some more testing next week.  For enabling CI I just removed the none and uncommented the lines.

![image](https://user-images.githubusercontent.com/127534928/229247591-9bca32ec-e22b-4c93-a624-2d601eddeaae.png)

**Reply by skeeler** [reply]
@theMK2000 did you push the committed changes back to your Azure DevOps repo? From the same error message shown, it looks like the pipeline is still using original code version.

**Reply by theMK2000** [reply]
Yes, I double-checked the code in ADO and it is updated and committed to the main branch.  You can see by the error above it is parsing the file path differently from before but there is still a higher folder structure that is passed with the /.  Though it seems to be passing the entire path.  

![image](https://user-images.githubusercontent.com/127534928/229555762-54d0427a-a1c4-4c62-8edd-7bcc4ca6a240.png)

 
subscriptions.yml

![image](https://user-images.githubusercontent.com/127534928/229554442-da790612-5be0-433a-9172-0705b0c8d19d.png)


trigger-subscriptions.yml

![image](https://user-images.githubusercontent.com/127534928/229554312-9d1904b3-0f61-4ddf-b00a-b42c8c6a6372.png)

**Reply by skeeler** [reply]
@theMK2000, you're missing a couple of "*" characters in the replace string for variable `$guids`. It should look like this (from my test environment):

![image](https://user-images.githubusercontent.com/755941/229617987-6b161f73-98fc-4f44-93c9-9a8e7a2d87ce.png)

Here is the code from earlier post above:

`$guids = $paths -replace '.*?([0-9a-f]{8}[-]?([0-9a-f]{4}[-]?){3}[0-9a-f]{12}).*', '$1'`

Try adding the missing asterisks and rerun.

**Reply by theMK2000** [reply]
@skeeler God thank you I copied and pasted not sure how that happened but it is now running! Thanks for your assistance with this.

**Reply by skeeler** [reply]
Great news and happy to help.
